### ISSUE TITLE
Jobs randomly stuck for a long time

### ISSUE THREAD

**Issue by jacek213** [mperham/sidekiq#4699]
Ruby version: 2.6.5 (jemalloc)
Rails version: 5.1.7
Sidekiq / Pro / Enterprise version(s): 5.2.9
Server: Redhat Enterprise 7.8

**sidekiq.yml**

```yaml
---
development:
  :verbose: true
  :concurrency: 5
production:
  :logfile: ./log/sidekiq.log
  :concurrency: 8
:pidfile: ./tmp/pids/sidekiq.pid
:queues:
  - default
  - asc_import
  - asc_import_equipment
  - mailers
  - renders_download
:schedule:
  clear_shrine_cache:
    cron: '00 01 * * *'
    class: ShrineCacheWorker
  check_stale_car_models:
    cron: '00 8,20 * * *'
    class: StaleCarModelsWorker
```

**Initializer**

```ruby
Sidekiq.configure_server do |config|
  config.redis = MockRedis.new if Rails.env.test?

  config.error_handlers << Proc.new do |ex, ctx_hash|
    ExceptionNotifier.notify_exception(ex)
  end
end

```

On my production server some jobs get stuck for amount of time which is usually in range of 1min to 60mins. From my observations, it happens to workers which make external HTTP calls. It's not an easy thing to debug because it happens randomly. However, it happens pretty often. Also, I have no idea what's actually going on during that time.

In the web panel I just see the job as **busy**. It always succeeds, however it takes a lot of time. External requests are made using Faraday (0.15.3).

![sidekiq-panel](https://user-images.githubusercontent.com/3914691/94021082-ec151780-fdb3-11ea-9015-931ccade3ecb.jpg)

**Example worker:**

```ruby
# frozen_string_literal: true

class WlsGenericLeadsWorker
  include Sidekiq::Worker
  sidekiq_options queue: :default, retry: 5, backtrace: true

  def perform(generic_lead_form_id)
    generic_lead_form = GenericLeadForm.find(generic_lead_form_id)
    payload = Wls::GenericLeadPayload.new(generic_lead_form)
    sender = Wls::Sender.new(payload)
    sender.call
  end
end

```

It just fetches a record from database by ID, prepares payload based on it's data and sends a POST request with Faraday. It happens for other workers as well.

**The log**

![log-sidekiq](https://user-images.githubusercontent.com/3914691/94022016-f257c380-fdb4-11ea-9576-fc384a971498.jpg)

As you can see, the job was started at 13:06 and the actual request was made at 13:21. What happened in between remains unknown.

**Appsignal panel**

![appsignal-screen](https://user-images.githubusercontent.com/3914691/94023971-20d69e00-fdb7-11ea-8037-89ed893cc321.jpg)

Here's also a screen from appsignal panel. As you can see CPU Usage is low and there's more than plenty of memory and swap.

**Timeline**

![image](https://user-images.githubusercontent.com/3914691/94025368-b6bef880-fdb8-11ea-9037-143642c57ab8.png)

The timeline indicates that `activerecord` and `net_http` were called at the very end and the job had been stuck at `sidekiq` itself for 15 minutes.

Puma server working on the same machine causes no problems and handles around 500 requests a minute.

Help would be more than welcome... I've been struggling with that for quite some time.

**Comment by mperham** [comment]
Use the TTIN signal, as noted on the Problems and Troubleshooting wiki page.

**Comment by jacek213** [comment]
@mperham thank you for a tip. 

I've called the TTIN signal on a main sidekiq process and it produce data into my logs. However, I'm not sure how to make a use of it. 

Please check the gist: https://gist.github.com/jacek213/6ce832e63a35d0603a4c0ccf52d464b5 containing log data.

**Comment by mperham** [comment]
You have multiple processor threads blocked at the exact same point, waiting on a postgresql query. For example:

```
2020-09-24T13:55:27.286Z 9675 TID-oumwvl8b7 WARN: Thread TID-oumsvzlhv processor
2020-09-24T13:55:27.287Z 9675 TID-oumwvl8b7 WARN: /home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:260:in `async_exec'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:260:in `block in active?'
/home/deploy/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/postgresql_adapter.rb:259:in `active?'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract_adapter.rb:444:in `verify!'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:810:in `block in checkout_and_verify'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:131:in `run_callbacks'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activesupport-5.1.7/lib/active_support/callbacks.rb:827:in `_run_checkout_callbacks'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:809:in `checkout_and_verify'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:502:in `checkout'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:376:in `connection'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:933:in `retrieve_connection'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_handling.rb:116:in `retrieve_connection'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/connection_handling.rb:88:in `connection'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/core.rb:301:in `cached_find_by_statement'
/home/deploy/myapp/shared/bundle/ruby/2.6.0/gems/activerecord-5.1.7/lib/active_record/core.rb:183:in `find'
/home/deploy/myapp/releases/20200924084752/app/workers/wls_generic_leads_worker.rb:8:in `perform'
```

I can't tell more than that but you might need an index on a table or have something else taking a table lock and blocking any other connection from accessing that table.

**Comment by jacek213** [comment]
@mperham thank you for your help.

In this case we only query by an indexed primary key (id).  The table is rarely used by the app in general - records are created in puma thread and the sidekiq worker is the only the place where they are being queried.

PostgreSQL max_connections value equals 100.

I have 8 sidekiq workers, 4 puma workers and 8 puma threads.

The database.yml pool size was set to 8, I've recently changed it to 16 for debugging purposes but it did not make any difference whatsoever.

Running 

```sql
select count(*) from pg_stat_activity where pid <> pg_backend_pid() and usename = current_user;
```
currently returns 34.

Do you have any idea what could possibly cause the lock?

**Comment by mperham** [comment]
I can't help debug database issues, that has nothing to do with Sidekiq -- it's between Rails, your app code and postgres. The TTIN signal helps you determine where your app code is blocked, it's up to you to diagnose.

**Comment by jacek213** [comment]
@mperham well, I'm not stating that Sidekiq is buggy or leaking connections by itself but from my point of view it does have to do with Sidekiq as I currently have 45 open database connections from which 32 come from Puma (which is expected - 4 workers x 8 threads) and the remaining 13 are Sidekiq connections. As I've stated I have 8 Sidekiq workers, so 13 - 8 = 5 connections are zombie from my understanding (please correct me if I'm wrong and if one worker opening more than one connection is not a leak). 

So my hypothesis is that if a new job receives one of the "broken" connections then it hangs until some timeout on Postgres' side happens.

Could wrapping my code inside `ActiveRecord::Base.connection_pool.with_connection` block be a potential solution?

**Comment by LyriD** [comment]
please help! i have the same trouble(maybe) in production. with payment(and another) jobs.
jobs freezing by 60 seconds. db pool is ok
it repeats every day
and passes as suddenly as it begins. for about an hour all jobs freezes for exactly one minute

sidekiq 6.0.5 
rails 6.0.2.1 
pg 1.2.2 

sidekiq.yml

production:
  :concurrency: 20
  :logfile: ./log/sidekiq.log
:timeout: 120
:queues:
  - default
  - mqtt
  - payments
  - mailers
  - email_stat
  - refresh
  - update_vending_problems
  - cache_vendings
:schedule:
  mqtt:
    cron: '*/4 * * * *' 
    class: UpdateCellsJob
  email_stat:
    cron: '0 2 * * *' 
    class: EmailStatJob
  refresh:
    cron: '*/15 * * * *'
    class: RefreshFaJob
  update_vending_problems:
    cron: '*/5 * * * *'
    class: UpdateVendingProblems
    enabled: true
  cache_vendings:
    cron: '*/6 * * * *' 
    class: CacheVendingsJob

this is kill --ttin pid output:

https://gist.github.com/LyriD/e1e3ae39f8fddfae80f8eff88038f4ac

Please! Help!

**Comment by mperham** [comment]
@LyriD That gist output makes it pretty clear who's at fault. Spend 30 minutes looking at that gist, the patterns therein and the code responsible. Reading a backtrace is a skill; you'll learn a lot.

**Comment by LyriD** [comment]
@mperham I see many different reasons, but I can not establish the truth, please help, problems in production right now.. and I have a little panic..

**Comment by LyriD** [comment]
@mperham this is trouble with elastic-apm timeouts?

**Comment by mperham** [comment]
Yes, every thread is stuck, waiting for Elastic to reply.

**Comment by LyriD** [comment]
@mperham very very big thanks ❤️

**Comment by nguyenngoc2505** [comment]
@mperham I've same problem. Could you help me check this?  This is TTIN signal log.
Thankyou so much!

https://gist.github.com/nguyenngoc2505/2d420e8514632406f48f2e3e378fdf77

**Comment by mperham** [comment]
@nguyenngoc2505 Look at the `processor` thread traces. Every one has the `mail` gem in it. See the note about `mail` here: https://github.com/mperham/sidekiq/wiki/Problems-and-Troubleshooting#use-only-thread-safe-libraries

**Comment by darveshsharma** [comment]
please help! files are stuck in jobs and not processing every process is being stuck working fine in local

**Comment by sergei-fedorov** [comment]
Hi @jacek213! We are experiencing a similar issue to what you described. Have you found a solution to your problem?

**Comment by cabello** [comment]
This happened to me recently for the first time ever, I have been using Sidekiq for 7 years now... I don't know I lost count. Thankful for this project existing. ❤️ 

Sidekiq versions 6.5.6 & 6.5.7.

Behaviour in my case:
* A specific worker class was failing to run for all jobs, it would go to the retry queue, the failure was due to an external API response having changed, we perform these API requests using Faraday (v.1.9.3)
* I don't even know how so many jobs of that failed worker were scheduled to begin with, I was expecting 1 job, but there were ~100
* Jobs were accumulating in the queue once that job failed enough times, latency would grow, our Rails Autoscale would kick in

<img width="1096" alt="CleanShot 2022-09-22 at 17 18 20@2x" src="https://user-images.githubusercontent.com/60936/191852871-3965a6bb-cca5-469e-946f-2f72655b0482.png">

Restarting the worker in Heroku would restart the processing of jobs until that failed worker was retried enough times, then Heroku would exist with code 22 and even though the Worker Dynos were up & running, they were not processing jobs anymore.

I changed that worker to early return in case of a bad API response, re-deployed and job processing resumed.

Here is the error my error tracker was receiving that made me a bit clueless:
<img width="1179" alt="CleanShot 2022-09-22 at 17 21 50@2x" src="https://user-images.githubusercontent.com/60936/191853427-302471ce-9f18-4ff2-ad3a-b9b930b56171.png">

I suspect my code was not thread-safe that caused Sidekiq job processing to accumulate open connections instead of returning them to the pool which resulted in starvation, the job processor needs a postgres connection to start with the job but none were available.

I am not certain if exit code 22 matches this page https://devcenter.heroku.com/articles/error-codes#h22-connection-limit-reached but that was the reasoning for this theory

<img width="762" alt="CleanShot 2022-09-22 at 17 24 56@2x" src="https://user-images.githubusercontent.com/60936/191853844-d877eb98-f157-4942-978a-b5b213fefd8d.png">
